cmd/maya-apiserver/spc-watcher/select_disk.go

```diff
@@ -0,0 +1,270 @@
+			nodeDiskMap[value.Labels[string(v1alpha1.HostNameCPK)]] = &nodeDisk{diskList: []string{value.Name}, diskCount: 1}
+			// If pool type is striped the node qualifies for pool creation hence pendingAllotment decremented.
+			if poolType == string(v1alpha1.PoolTypeStripedCPK) {
+				pendingAllotment--
+			}
+		} else {
+			// Entry to this block means the hostname was already mapped and it has more than one disk and at least two disks.
+			nodeDisk := nodeDiskMap[value.Labels[string(v1alpha1.HostNameCPK)]]
+			// Increment the disk count
+			nodeDisk.diskCount++
+			// Add the current disk to the diskList for this node.
+			nodeDisk.diskList = append(nodeDisk.diskList, value.Name)
+			// If pool type is mirrored the node qualifies for pool creation hence pendingAllotment decremented.
+			if poolType == string(v1alpha1.PoolTypeMirroredCPK) {
+				if nodeDisk.diskCount == int(v1alpha1.MirroredDiskCountCPK) {
+					pendingAllotment--
+				}
+			}
+		}
+
+	}
+	return nil, nodeDiskMap, pendingAllotment
+}
+
+// diskSelector is the function that will select the required number of disks from qualified nodes
+// so as to provision storagepool
+
+func diskSelector(nodeDiskMap map[string]*nodeDisk, poolType string) []string {
+
+	// selectedDisk will hold a list of disk that will be used to provision storage pool, after a
+	// minimum number of node qualifies
+	var selectedDisk []string
+
+	// requiredDiskCount will hold the required number of disk that should be selcted from a qualified
+	// node for specific pool type
+	var requiredDiskCount int
+
+	// Range over the nodeDiskMap map to get the list of disks
+	for _, val := range nodeDiskMap {
+		// If pool type is striped, 1 disk should be selected
+		if poolType == string(v1alpha1.PoolTypeStripedCPK) {
+			requiredDiskCount = int(v1alpha1.StripedDiskCountCPK)
+		}
+		// If pool type is mirrored, 2 disks should be selected
+		if poolType == string(v1alpha1.PoolTypeMirroredCPK) {
+			requiredDiskCount = int(v1alpha1.MirroredDiskCountCPK)
+			// If the current disk count on the node is less than the required disks
+			// then this is a dirty node and it will not qualify.
+			if len(val.diskList) < requiredDiskCount {
+				continue
+			}
+		}
+		// Select the required disk from qualified nodes.
+		for i := 0; i < requiredDiskCount; i++ {
+			selectedDisk = append(selectedDisk, val.diskList[i])
+		}
+	}
+	return selectedDisk
+}
+
+// diskFilterConstraint will form labels that will be used to filter disks
+
+func diskFilterConstraint(diskType string) string {
+	var label string
+	if diskType == string(v1alpha1.TypeSparseCPK) {
+		label = string(v1alpha1.DiskTypeCPK) + "=" + string(v1alpha1.TypeSparseCPK)
+	} else {
+		label = string(v1alpha1.DiskTypeCPK) + "=" + string(v1alpha1.TypeDiskCPK)
+	}
+	return label
+}
+
+// form usedDisk map that will hold the list of all used disks
+
+func (k *clientSet) getUsedDiskMap() (error, map[string]int) {
```

Comment:
Why this has to be part of clientSet ?

Reply:
It needs clientSet to query for SP objects.